File: w1/time_conversion.ts
/**
 * Time Conversion - Challenge 3 - Week 1
 * https://www.hackerrank.com/challenges/three-month-preparation-kit-time-conversion/problem
 * 
 * @param s - A time string in the form 12:40:22AM
 * @returns - A 'military' formatted time 00:40:22
 * 
 * @example
 * timeConversion("12:40:22AM") // "00:40:22"
 * timeConversion("07:05:45PM") // "19:05:45"
 */
export function timeConversion(s:string): string {
    // get into parts
    const [, hourStr, minStr, secStr, period] =
    s.match(/^(\d{2}):(\d{2}):(\d{2})(AM|PM)$/)!;

    // hour is the target of our meddling.
    let hour = parseInt(hourStr);

    // if period is PM add 12 to hour.
    if (period === "PM" && hour !== 12) hour += 12;

    // if period is AM and hour is twelve, then it moves to zero.
    if (period === "AM" && hour === 12) hour = 0;

    // format and return
    return `${hour.toString().padStart(2, "0")}:${minStr}:${secStr}`;
}

File: w1/time_conversion.ex
defmodule HackerRank.W1.TimeConversion do
  @moduledoc """
  Time Conversion – Challenge 3 – Week 1
  <https://www.hackerrank.com/challenges/three-month-preparation-kit-time-conversion/problem>
  """

  @doc """
  Converts 12-hour AM/PM time format to 24-hour military time.

  ## Examples

      iex> challenge("07:05:45PM")
      "19:05:45"

  ## Constraints

  - Input format: `"hh:mm:ssAM"` or `"hh:mm:ssPM"`
  - Always returns a string in 24-hour format
  """
  @spec challenge(String.t()) :: String.t()
  def challenge(s) do
    # Dates / Times are not Elixis strong suit
    # However, binary string pattern matching is so let's do that then
    [hh, mm, ss_am_pm] = String.split(s, ":")

    # now we need am / pm
    <<ss::binary-size(2), period::binary>> = ss_am_pm

    # now parse into the correct types
    hour = parse_hour(hh, period)
    minute = String.to_integer(mm)
    second = String.to_integer(ss)

    # run everything through pad and return with Enum.map_join
    Enum.map_join([hour, minute, second], ":", &pad2/1)
    end

  defp parse_hour("12", "AM"), do: 0
  defp parse_hour(h, "AM"), do: String.to_integer(h)
  defp parse_hour("12", "PM"), do: 12
  defp parse_hour(h, "PM"), do: String.to_integer(h) + 12

  defp pad2(number) do
    :io_lib.format("~2..0B", [number]) |> to_string()
  end
end
